internal/notifier/pagerduty.go

```diff
@@ -0,0 +1,132 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package notifier
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/PagerDuty/go-pagerduty"
+	eventv1 "github.com/fluxcd/pkg/apis/event/v1beta1"
+	"github.com/fluxcd/pkg/apis/meta"
+	"k8s.io/utils/strings/slices"
+)
+
+var (
+	validEndpoints = []string{
+		"events.pagerduty.com",
+		"events.eu.pagerduty.com",
+	}
+)
+
+type PagerDuty struct {
+	Client     *pagerduty.Client
+	RoutingKey string
+}
+
+func NewPagerDuty(endpoint string, routingKey string) (*PagerDuty, error) {
+	pURL, err := url.ParseRequestURI(endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("invalid PagerDuty endpoint URL %s: '%w'", endpoint, err)
+	}
+	pHostname := pURL.Hostname()
+	if !slices.Contains(validEndpoints, pHostname) {
+		return nil, fmt.Errorf("invalid PagerDuty endpoint hostname %s: should be one of: %v", pHostname, validEndpoints)
+	}
+	opts := []pagerduty.ClientOptions{
+		pagerduty.WithV2EventsAPIEndpoint("https://" + pHostname),
+	}
+	// No token needed for our usage of the V2 events API
+	client := pagerduty.NewClient("", opts...)
+	return &PagerDuty{
+		Client:     client,
+		RoutingKey: routingKey,
+	}, nil
+}
+
+func (p *PagerDuty) Post(ctx context.Context, event eventv1.Event) error {
+	// skip commit status updates and progressing events (we want success or failure)
+	if event.HasMetadata(eventv1.MetaCommitStatusKey, eventv1.MetaCommitStatusUpdateValue) || event.HasReason(meta.ProgressingReason) {
+		return nil
+	}
+	e := toPagerDutyV2Event(event, p.RoutingKey)
+	_, err := p.Client.ManageEventWithContext(ctx, &e)
+	if err != nil {
+		return fmt.Errorf("failed to manage PagerDuty event: %w", err)
+	}
+	// Send a change event for info events
+	if event.Severity == eventv1.EventSeverityInfo {
+		ce := toPagerDutyChangeEvent(event, p.RoutingKey)
+		_, err := p.Client.CreateChangeEventWithContext(ctx, ce)
+		if err != nil {
+			return fmt.Errorf("failed to create PagerDuty change event: %w", err)
+		}
+	}
+	return nil
+}
+
+func toPagerDutyV2Event(event eventv1.Event, routingKey string) pagerduty.V2Event {
+	name, desc := formatNameAndDescription(event)
+	// Send resolve just in case an existing incident is open
+	e := pagerduty.V2Event{
+		RoutingKey: routingKey,
+		Action:     "resolve",
+		DedupKey:   string(event.InvolvedObject.UID),
+	}
+	// Trigger an incident for errors
+	if event.Severity == eventv1.EventSeverityError {
+		e.Action = "trigger"
+		e.Payload = &pagerduty.V2Payload{
+			Summary:   desc + ": " + name,
+			Source:    "Flux " + event.ReportingController,
+			Severity:  toPagerDutySeverity(event.Severity),
+			Timestamp: event.Timestamp.Format(time.RFC3339),
+			Component: event.InvolvedObject.Name,
+			Group:     event.InvolvedObject.Kind,
+			Details: map[string]interface{}{
```

Comment:
We need to add the event metadata to the details, the metadata can contain various infos such as source revision, cluster name, env, region, etc